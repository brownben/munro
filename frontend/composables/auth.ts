import type { Ref } from 'vue'

interface User {
  idToken: string
  displayName?: string
}

const useUserState = () => useState<User | null>('user')

export const useUserCookie = () =>
  useCookie<User | null>('user', { maxAge: 3600, sameSite: true })

export const loginRequest = async (
  email: Ref<string>,
  password: Ref<string>
) => {
  const { idToken, displayName } = await $fetch<User>(
    'https://identitytoolkit.googleapis.com/v1/accounts:signInWithPassword?key=AIzaSyAQriY0O2Atf-En8yKMXNs5TIRCglWuAbQ',
    {
      body: {
        email: email.value,
        password: password.value,
        returnSecureToken: true,
      },
      method: 'POST',
    }
  )
  useUserCookie().value = { idToken, displayName }
  useUserState().value = { idToken, displayName }
}

export const logoutRequest = () => {
  useUserCookie().value = null
  useUserState().value = null
}

export const useUser = () => {
  const cookie = useUserCookie()
  const user = useUserState()

  if (cookie.value && !user.value) user.value = cookie.value

  return user
}

export const useLoggedIn = () => {
  const user = useUser()

  return computed(() => !!user.value?.idToken)
}

export const redirect = async (path: string) => {
  const app = useNuxtApp()
  if (app.ssrContext) {
    app.ssrContext.res.writeHead(302, { Location: path })

    return app.ssrContext.res.end()
  }

  return await useRouter().push(path)
}

export const useRequireAuthentication = async () => {
  const user = useLoggedIn()

  if (!user.value) return redirect('/login')
}
