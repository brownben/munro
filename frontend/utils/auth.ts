import type { Ref } from 'vue'

interface User {
  idToken: string
  displayName?: string
}

const useUserState = () => useState<User | null>('user')

export const useUserCookie = () =>
  useCookie<User | null>('user', { maxAge: 3600, sameSite: true })

export const loginRequest = async (
  email: Ref<string>,
  password: Ref<string>,
) => {
  const { idToken, displayName } = await $fetch<User>(
    'https://identitytoolkit.googleapis.com/v1/accounts:signInWithPassword?key=AIzaSyAQriY0O2Atf-En8yKMXNs5TIRCglWuAbQ',
    {
      body: {
        email: email.value,
        password: password.value,
        returnSecureToken: true,
      },
      method: 'POST',
    },
  )
  useUserCookie().value = { idToken, displayName }
  useUserState().value = { idToken, displayName }
}

export const logoutRequest = () => {
  useUserCookie().value = null
  useUserState().value = null
}

export const useUser = () => {
  const cookie = useUserCookie()
  const user = useUserState()

  if (cookie.value && !user.value) user.value = cookie.value

  return user
}

export const useLoggedIn = () => {
  const user = useUser()

  return computed(() => !!user.value?.idToken)
}

export const requireLogin = () => {
  const user = useUser()
  const route = useRoute()

  if (!user.value?.idToken) {
    return redirect(`/login?redirect=${route.fullPath}`)
  }
}

export const redirect = async (path: string) => {
  return navigateTo(path, { replace: true })
}
