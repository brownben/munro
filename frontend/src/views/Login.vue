<!--
  Login

  The login page to gain access to admin functions
-->

<template>
  <div>
    <h1>Admin Login</h1>
    <form @submit.prevent="sendLoginRequest()">
      <label>Email Address:</label>
      <input v-model="username" type="text">
      <label>Password:</label>
      <input v-model="password" type="password">
      <button>Login</button>
    </form>
    <div id="link-to-upload">
      <h3>Looking for Results Upload?</h3>
      <p>
        Results are uploaded
        <router-link to="/upload">Here</router-link>
      </p>
    </div>
  </div>
</template>

<script>
import auth from '@/authentication'

export default {
  data () {
    return {
      username: '',
      password: '',
    }
  },

  mounted: function () {
    if (this.$auth.user) this.$router.push('/')
    this.blankFields()
    if (this.$route.query.redirect) this.$messages.addMessage('Please Login to Access that Page')
  },

  methods: {
    blankFields: function () {
      this.username = ''
      this.password = ''
    },

    validateLogin: function () {
      return (this.username !== '' && this.password !== '')
    },

    sendLoginRequest: function () {
      if (this.validateLogin()) {
        return auth.login(this.username, this.password)
          .then(response => {
            if (response) this.$router.replace(this.$route.query.redirect || '/')
            this.$messages.addMessage('Hello')
            this.blankFields()
          })
          .catch(() => this.$messages.addMessage('Error: Problem Logging In - Please Try Again'))
      }
      else this.$messages.addMessage('Problem: Username or Password were left Blank')
    },
  },
}

</script>
<style lang="stylus" scoped>
@import '../assets/styles/helpers.styl'
@import '../assets/styles/inputs.styl'

#router-view
  margin-left: 5%
  padding-top: 1rem
  width: 90%

#container
  margin-left: 5%
  width: 90%

h1
  margin-bottom: 0.5rem

button
  margin-top: 0.5rem

input
  width: 100% !important

#link-to-upload
  margin: 0
  padding: 1rem 0
  width: 100%

  p, h3
    display: inline-block

.fade-enter-active, .fade-leave-active
  transition: opacity 0.5s

.fade-enter, .fade-leave-to
  opacity: 0
</style>
