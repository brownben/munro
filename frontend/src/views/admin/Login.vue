<!--
  Login

  The login page to gain access to admin functions
-->

<template>
  <Layout title="Admin Login">
    <Meta
      :head="{
        meta: { name: 'robots', content: 'noindex' },
      }"
      title="Munro - Login"
      description
    />
    <form class="col-span-2" @submit.prevent="sendLoginRequest">
      <TextInput v-model="username" label="Email Address:" type="email" />
      <TextInput
        v-model="password"
        label="Password:"
        type="password"
        class="mt-4"
      />

      <button class="mt-6 button-lg">
        Login
      </button>
    </form>

    <div class="col-span-2 p-4 card card-color">
      <p>
        <b
          class="block mb-2 text-xl font-heading md:m-0 md:text-lg md:inline-block md:mr-4"
        >
          Looking for Results Upload?
        </b>
        You don't have to login, just visit the
        <router-link to="/upload" class="link">
          upload page
        </router-link>
      </p>
    </div>
  </Layout>
</template>

<script>
import auth from '/@/authentication'

import Layout from '/@/components/Layout.vue'
import TextInput from '/@/components/inputs/TextInput.vue'

export default {
  components: {
    Layout,
    TextInput,
  },

  data() {
    return {
      username: '',
      password: '',
    }
  },

  mounted: function () {
    if (this.$auth.user) {
      this.$store.dispatch('createMessage', 'You Are Already Logged In')
      this.$router.push('/')
    }
    this.blankFields()
    if (this.$route.query.redirect)
      this.$store.dispatch('createMessage', 'Please Login to Access that Page')
  },

  methods: {
    blankFields: function () {
      this.username = ''
      this.password = ''
    },

    validateLogin: function () {
      return this.username !== '' && this.password !== ''
    },

    sendLoginRequest: function () {
      if (this.validateLogin()) {
        return auth
          .login(this.username, this.password)
          .then((response) => {
            if (response)
              this.$router.replace(this.$route.query.redirect || '/')
            this.$store.dispatch(
              'createMessage',
              `Hello ${this.$auth.user.displayName || 'Admin'}`
            )
            this.blankFields()
          })
          .catch(() =>
            this.$store.dispatch(
              'createMessage',
              'Error: Problem Logging In - Please Try Again'
            )
          )
      } else
        this.$store.dispatch(
          'createMessage',
          'Problem: Username or Password were left Blank'
        )
    },
  },
}
</script>
