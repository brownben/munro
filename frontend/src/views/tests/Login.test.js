/*
  Login View Unit Tests
*/

import { mount } from '@vue/test-utils'
import Login from '@/views/Login'
import auth from '@/authentication'

auth.login = jest.fn()
auth.logout = jest.fn()

beforeEach(() => jest.clearAllMocks())

test('Is a Vue Instance', () => {
  const wrapper = mount(Login, {
    mocks: {
      $route: { query: { redirect: false } },
      $auth: { user: false },
    },
    stubs: ['router-link', 'vue-headful'],
  })
  expect(wrapper.isVueInstance()).toBeTruthy()
})

test('Renders Correctly', () => {
  const wrapper = mount(Login, {
    mocks: {
      $route: { query: { redirect: false } },
      $auth: { user: false },
    },
    stubs: ['router-link', 'vue-headful'],
  })
  expect(wrapper.element).toMatchSnapshot()
})

test('Show Message on When Redirected', () => {
  const mockAddMessageFunction = jest.fn()
  mount(Login, {
    mocks: {
      $route: { query: { redirect: '/leagues' } },
      $messages: { addMessage: mockAddMessageFunction },
      $auth: { user: false },
    },
    stubs: ['router-link', 'vue-headful'],
  })
  expect(mockAddMessageFunction).toHaveBeenCalledTimes(1)
  expect(mockAddMessageFunction).toHaveBeenLastCalledWith(
    'Please Login to Access that Page'
  )
})

test('Blank Fields Works', () => {
  const wrapper = mount(Login, {
    mocks: {
      $route: { query: { redirect: false } },
      $auth: { user: false },
    },
    stubs: ['router-link', 'vue-headful'],
  })
  wrapper.setData({
    username: 'username',
    password: 'password',
  })
  expect(wrapper.vm.username).toBe('username')
  expect(wrapper.vm.password).toBe('password')
  wrapper.vm.blankFields()
  expect(wrapper.vm.username).toBe('')
  expect(wrapper.vm.password).toBe('')
})

test('Check Validation Works', () => {
  const wrapper = mount(Login, {
    mocks: {
      $route: { query: { redirect: false } },
      $auth: { user: false },
    },
    stubs: ['router-link', 'vue-headful'],
  })
  wrapper.setData({ username: '', password: '' })
  expect(wrapper.vm.validateLogin()).toBeFalsy()

  wrapper.setData({ username: 'value', password: '' })
  expect(wrapper.vm.validateLogin()).toBeFalsy()

  wrapper.setData({ username: '', password: 'value' })
  expect(wrapper.vm.validateLogin()).toBeFalsy()

  wrapper.setData({ username: 'A Value', password: 'Another Value' })
  expect(wrapper.vm.validateLogin()).toBeTruthy()

  wrapper.setData({
    username: 'username',
    password: 'password',
  })
  expect(wrapper.vm.validateLogin()).toBeTruthy()
})

test('Send Login Function - Successful', async () => {
  const mockAddMessageFunction = jest.fn()
  auth.login.mockResolvedValue({ data: { loggedIn: true, message: 'Hello' } })
  const wrapper = mount(Login, {
    mocks: {
      $route: { query: { redirect: false } },
      $router: { replace: jest.fn() },
      $messages: { addMessage: mockAddMessageFunction },
      $auth: { user: false },
    },
    stubs: ['router-link', 'vue-headful'],
  })
  const mockBlankFields = jest.fn()
  wrapper.setMethods({ blankFields: mockBlankFields })
  // Calls Auth Login Functions
  wrapper.setData({ username: 'username', password: 'password' })
  auth.login.mockResolvedValue({ data: { loggedIn: true, message: 'Hello' } })

  await wrapper.vm.sendLoginRequest()
  expect(auth.login).toHaveBeenCalledTimes(1)
  expect(auth.login).toHaveBeenLastCalledWith('username', 'password')

  // Correct Actions On Completion
  expect(mockBlankFields).toHaveBeenCalledTimes(1)
  expect(mockAddMessageFunction).toHaveBeenCalledTimes(1)
  expect(mockAddMessageFunction).toHaveBeenLastCalledWith('Hello Admin')
})

test('Send Login Function - Validation Fails', () => {
  const mockAddMessageFunction = jest.fn()
  const wrapper = mount(Login, {
    mocks: {
      $route: { query: { redirect: false } },
      $messages: { addMessage: mockAddMessageFunction },
      $auth: { user: false },
    },
    stubs: ['router-link', 'vue-headful'],
  })
  auth.login.mockRejectedValue()
  const blankFieldsSpy = jest.spyOn(wrapper.vm, 'blankFields')
  wrapper.setData({ username: '', password: '' })
  wrapper.vm.sendLoginRequest()
  expect(blankFieldsSpy).toHaveBeenCalledTimes(0)
  expect(mockAddMessageFunction).toHaveBeenCalledTimes(1)
  expect(mockAddMessageFunction).toHaveBeenLastCalledWith(
    'Problem: Username or Password were left Blank'
  )
})

test('Send Login Function - Error', () => {
  const mockAddMessageFunction = jest.fn()
  auth.login.mockRejectedValue()
  const wrapper = mount(Login, {
    mocks: {
      $route: { query: { redirect: false } },
      $messages: { addMessage: mockAddMessageFunction },
      $auth: { user: false },
    },
    stubs: ['router-link', 'vue-headful'],
  })
  const blankFieldsSpy = jest.spyOn(wrapper.vm, 'blankFields')

  // Calls Auth Login Functions
  wrapper.setData({ username: 'username', password: 'password' })
  return wrapper.vm.sendLoginRequest().then(() => {
    expect(auth.login).toHaveBeenCalledTimes(1)
    expect(auth.login).toHaveBeenLastCalledWith('username', 'password')

    // Expect Correct Function on Response
    expect(blankFieldsSpy).toHaveBeenCalledTimes(0)
    expect(mockAddMessageFunction).toHaveBeenCalledTimes(1)
    expect(mockAddMessageFunction).toHaveBeenLastCalledWith(
      'Error: Problem Logging In - Please Try Again'
    )
  })
})

test('Redirect on Successful Login', async () => {
  const mockAddMessageFunction = jest.fn()
  auth.login.mockResolvedValue({
    data: { loggedIn: true, message: 'Hello User' },
  })
  const wrapper = mount(Login, {
    mocks: {
      $route: { query: { redirect: '/other' } },
      $router: { replace: jest.fn() },
      $messages: { addMessage: mockAddMessageFunction },
      $auth: { user: false },
    },
    stubs: ['router-link', 'vue-headful'],
  })

  // Calls Auth Login Functions
  wrapper.setData({ username: 'username', password: 'password' })
  await wrapper.vm.sendLoginRequest()

  expect(wrapper.vm.$router.replace).toHaveBeenCalledTimes(1)
  expect(wrapper.vm.$router.replace).toHaveBeenLastCalledWith('/other')
})

test('Redirect on Successful Login - If No Path Specified', async () => {
  const mockAddMessageFunction = jest.fn()
  auth.login.mockResolvedValue({
    data: { loggedIn: true, message: 'Hello User' },
  })
  const wrapper = mount(Login, {
    mocks: {
      $route: { query: { redirect: false } },
      $router: { replace: jest.fn() },
      $messages: { addMessage: mockAddMessageFunction },
      $auth: { user: false },
    },
    stubs: ['router-link', 'vue-headful'],
  })

  // Calls Auth Login Functions
  wrapper.setData({ username: 'username', password: 'password' })
  await wrapper.vm.sendLoginRequest()
  expect(wrapper.vm.$router.replace).toHaveBeenCalledTimes(1)
  expect(wrapper.vm.$router.replace).toHaveBeenLastCalledWith('/')
})

test("Don't Redirect on Failed Login", async () => {
  const mockAddMessageFunction = jest.fn()
  auth.login.mockResolvedValue(false)
  const wrapper = mount(Login, {
    mocks: {
      $route: { query: { redirect: '/other' } },
      $router: { replace: jest.fn() },
      $messages: { addMessage: mockAddMessageFunction },
      $auth: { user: false },
    },
    stubs: ['router-link', 'vue-headful'],
  })

  // Calls Auth Login Functions
  wrapper.setData({ username: 'username', password: 'password' })
  await wrapper.vm.sendLoginRequest()
  expect(wrapper.vm.$router.replace).toHaveBeenCalledTimes(0)
})

test('Redirect if Already Logged In', () => {
  const wrapper = mount(Login, {
    mocks: {
      $route: { query: { redirect: false } },
      $router: { push: jest.fn() },
      $messages: { addMessage: jest.fn() },
      $auth: { user: true },
    },
    stubs: ['router-link', 'vue-headful'],
  })

  // Calls Auth Login Functions
  expect(wrapper.vm.$router.push).toHaveBeenCalledTimes(1)
  expect(wrapper.vm.$router.push).toHaveBeenLastCalledWith('/')
})
