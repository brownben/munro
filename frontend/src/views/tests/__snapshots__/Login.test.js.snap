// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders Correctly 1`] = `
<div>
  <h1>
    Admin Login
  </h1>
   
  <form>
    <label>
      Email Address:
    </label>
     
    <input
      type="text"
    />
     
    <label>
      Password:
    </label>
     
    <input
      type="password"
    />
     
    <button>
      Login
    </button>
  </form>
   
  <div
    id="link-to-upload"
  >
    <h3>
      Looking for Results Upload?
    </h3>
     
    <p>
      
      Results are uploaded
      
      <router-link-stub
        to="/upload"
      >
        Here
      </router-link-stub>
    </p>
  </div>
</div>
`;
